2016/proposals/narayan-aditya-visualizer.md

```diff
@@ -0,0 +1,244 @@
+workshops fill in different surveys to allow the organization hosts
+to gather insight into the attendees computing background while
+the administrators collect long-term feedback.
+These surveys, which are hosted at [SurveyMonkey], are being
+analyzed using the SurveyMonkey web dashboard. Filtering the responses
+by a particular workshop can be cumbersome for the organization hosts
+and administrators as it requires a lot of clicks.
+
+The aim of this project is to build a web application to collect
+and visualize survey responses for every workshop, while providing
+a better experience for the administrators and organization hosts.
+
+## Technical Details
+
+The *Survey Visualizer* application stack can be divided into:
+
+ * Backend  : [Django Rest Framework]
+ * Scheduler: [Cron]
+ * Database : [PostgreSQL]
+ * Frontend : [Charts.js] and [Bootstrap]
+
+### Backend - Django Application
+
+The survey visualizer application would be made using
+[Django Rest Framework] (henceforth called DRF) over [Django]
+to allow easier integration with [AMY] when required.
+
+The tentative API structure exposed by DRF is:
+ * GET `/api/workshop/`  
+   Lists all workshops that have surveys collecting responses.
+ * GET `/api/workshop/<workshop_id>/`  
+   Returns details of a particular workshop.
+ * GET `/api/workshop/<workshop_id>/survey/`  
+   Lists all the surveys whose responses are being collected.
+   Each survey has its corresponding `survey_id`.
+ * POST `/api/workshop/<workshop_id>/survey/refetch`  
+   Repopulates the list of surveys whose response are being
+   collected.
+ * GET `/api/workshop/<workshop_id>/survey/<survey_id>`  
+   Returns the responses for the particular survey, listed
+   according to the questions contained in the survey.
+ * POST `/api/workshop/<workshop_id>/survey/<survey_id>/refetch`  
+   Instructs the Django application to fetch the responses
+   that the survey has received since the last time the
+   SurveyMonkey API was hit.
+
+The underlying Django app will perform these tasks:
+ * Polling SurveyMonkey API management command:
+   This management command would populate the database with survey
+   responses fetched from the SurveyMonkey API.
+   The number of survey responses obtained from the [get_response_counts]
+   API call will be compared to the number of survey responses
+   saved in the database. The remaining responses will be fetched
+   using the [get_respondent_list] and [get_responses] API calls.
+ * Segregating the survey responses according to the answer to the
+   question: *Which workshop did you attend?* or the custom variable
+   `workshop_id`.
+ * Rendering and serving the templates.
+
+### Scheduler - Cron
+
+[Cron] will be used as a job scheduler to periodically run
+the *Polling SM API* management command at fixed time intervals.
+The interval duration would be dynamically determined by the
+number of API calls left and the time of the day.
```

Comment:
We can only fetch 100 responses in one get_responses API call.
So, if we assume that we have received `x` survey responses for a survey in the last month, we need `2*x/100` API calls. (one each for get_respondent_list and get_responses).

For the [sample django app](http://infinite-falls-5979.herokuapp.com/), two API calls and one iteration over 6 responses cost 1.3 seconds (no caching though). While developing the app locally, it shoots to >5 sec. This was the reason why I choose to create an API and call it from the frontend. A traditional request-response cycle would seem sluggish.

Another benefit of keeping the SM API polling independent of the usual request-response cycle is that we would